### PR TITLE
Pull-Request: set up associations

### PR DESCRIPTION
**Pull request by Zven94**
### **School Library**

In this project, all classes were creating for the application in order to add associations between them.

Files added:

- book.rb
- classroom.rb
- rental.rb

### CODE REVIEW

File: student.rb [hunk 1]
```diff
@@ -1,11 +1,12 @@
...
+  # define getter&setter for @name and @classroom
+  attr_accessor :name, :classroom
+
   # Initializes a Student object with the given classroom
-  def initialize(age, classroom, name = 'Unknown', parent_permission: true)
+  def initialize(age, _classroom, name = 'Unknown', parent_permission: true)
     super(name, age, parent_permission)
-    @classroom = classroom
+    @classroom = nil
   end
```
Comment (topeogunleye) [inline student.rb:18]: Hi, you have done a great job, thank you for submitting your code. I have reviewed it and found some areas that need improvement.

- Your code does not meet the project requirement of setting the classroom for a student and adding it to the classroom's students. You are missing the `classroom=` method that serves as the 'belongs-to' side of the relationship between a student and a classroom. You can read more about this concept in this [article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#both-ways-relationships).


## Here is a possible way to refactor your code:

```.rb
class Student < Person
  # define getter for @classroom
  attr_reader :classroom

  # Initializes a Student object with the given classroom
  def initialize(age, classroom, name = 'Unknown', parent_permission: true)
    super(name, age, parent_permission)
    self.classroom = classroom # use the setter method to assign the classroom and add the student to it
  end

  # Defines a setter method for @classroom that also adds the student to the classroom's students
  def classroom=(classroom)
    @classroom = classroom
    classroom.students.push(self) unless classroom.students.include?(self)
  end

end
```

File: student.rb [hunk 2]
```diff
@@ -1,11 +1,12 @@
-require_relative 'person'
-
 # creating a Person subclass: student
 class Student < Person
+  # define getter&setter for @name and @classroom
+  attr_accessor :name, :classroom
```
Comment (topeogunleye) [inline student.rb]: - Kindly note that the `attr_accessor :name, :classroom` is not a method that serves as the 'belongs-to' side of the relationship between a student and a classroom. The `attr_accessor` is a shortcut method that creates both a reader and a writer method for the given instance variables¹. For example, `attr_accessor :name` creates two methods: `name` and `name=` that allow you to get and set the value of `@name`.

However, the project requirement is to create a custom setter method for `@classroom` that also adds the student to the classroom's students array. This is not possible with `attr_accessor`, as it only creates a simple assignment method. You need to define your own `classroom=` method that takes care of both actions, as shown in the project instructions².

Source: Conversation with Bing, 0[1](https://stackoverflow.com/questions/4370960/what-is-attr-accessor-in-ruby)/09/2023
(1) What is attr_accessor in Ruby? - Stack Overflow. https://stackoverflow.com/questions/4370960/what-is-attr-accessor-in-ruby.
([2](https://www.rubyguides.com/2018/11/attr_accessor/)) How to Use attr_accessor, attr_writer & attr_reader - RubyGuides. https://www.rubyguides.com/2018/11/attr_accessor/.
(3) What is attr_accessor in Ruby? - Educative. https://www.educative.io/answers/what-is-attraccessor-in-ruby.
(4) Ruby – attr_accessor, attr_writer, and attr_reader – bootrAils. https://bootrails.com/blog/ruby-attr-accessor-attr-writer-attr-reader/.

File: book.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# define Book class
+class Book
+  attr_accessor :title, :author
+  attr_reader :rental
+
+  def initialize(title, author)
+    @title = title
+    @author = author
+  end
+end
```
Comment (topeogunleye) [inline book.rb]: Great job on implementing the `Book` class! It looks like you've made progress towards meeting the project requirements for creating a many-to-many relationship between `Person` and `Book` using the intermediate class `Rental`. Your code seems to be on the right track, but there are a couple of points to consider for further improvement and alignment with the specified requirements:

1. **Relationship Setup:**
   Your `Book` class includes the necessary attributes `title`, `author`, and `rental`. However, to fully establish the many-to-many relationship, it's essential to maintain a collection of rentals associated with each book. In your `initialize` method, consider initializing the `@rentals` instance variable as an empty array to store rentals related to this book.

2. **Adding Rentals:**
   The requirement mentions that there should be a method for adding rentals to a book. You can create a method like `add_rental(person, date)` in your `Book` class, where you use the `Rental` class to create a new rental instance and add it to the `@rentals` array. This way, you establish the 'has-many' association between books and rentals.

Here's an example of how you could implement the `add_rental` method:
```ruby
class Book
  # ... (other attributes)

  def initialize(title, author)
    # ... (initialize other attributes)
    @rentals = []  # Initialize rentals as an empty array
  end

  def add_rental(person, date)
    rental = Rental.new(date, self, person)
    @rentals << rental  # Add the new rental to the rentals array
  end
end
```

By making these adjustments, your code will effectively create and manage the many-to-many relationship between `Person` and `Book` through the `Rental` intermediate class, as required.

Remember, you're making good progress! Feel free to refer to the provided resources for additional guidance on implementing the necessary methods and relationships. Keep up the good work!

Best regards,
Temitope
Code Reviewer
01/09/2023

---

File: person.rb [hunk 4]
```diff
@@ -21,6 +21,7 @@ def correct_name
 
   # define getter&setter for @name & @age
   attr_accessor :name, :age
+  attr_reader :rental
```
Comment (topeogunleye) [inline person.rb]: Great job on implementing the `rental` getter method! It looks like you've made progress towards meeting the project requirements for creating the `many-to-many` (also `has-many-through`) relationship between `Person` and `Book` using the intermediate class `Rental`. Your code seems to be on the right track, but there are a couple of points to consider for further improvement and alignment with the specified requirements:


1. **Adding Rentals:**
   The requirement mentions that there should be a method for adding rentals to a person. You can create a method like `add_rental(book, date)` in your `Person` class.


Remember, you're making good progress! Feel free to refer to the provided resources for additional guidance on implementing the necessary methods and relationships. Keep up the good work!

Best regards,
Temitope
Code Reviewer
01/09/2023

---

Comment (topeogunleye) [inline person.rb]: - [ ] As written in the project requirements, the ` *Getters and Setters for @rentals should be present`. please make changes. Kindly see this link here for helpful [info](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_principles_examples.md#useful-shortcuts-attr_reader-attr_writer-and-attr_accessor)

**Review by topeogunleye — CHANGES_REQUESTED** [review]
Hi @Zven94, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

Try-Again

> ## (Highlights) Good Points: 💪 
> * `Classroom` class implemented correctly. :heavy_check_mark:
> * `Book` class implemented correctly. :heavy_check_mark:
> * Rental class implemented correctly :heavy_check_mark:

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- x
- y
- z

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @topeogunleye in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: student.rb (hunk 1 above)
Comment (Zven94) [reply student.rb:18]: Hi @topeogunleye ! I want to thank you for your time reviewing my project. I'm going to implement all the changes about it.
By the way. the play_hook method was a requirement from the previous instances of the project. I don't think i have to remove it.

File: student.rb (hunk 2 above)
Comment (Zven94) [reply student.rb]: Thank you for the links, some of those articles i have already read it, but the problem i had was trying to understand how to create a method that serves as the 'belongs-to' side of the relationship

**Review by LionRouge1 — APPROVED** [review]
Hi @Zven94, 

# STATUS: APPROVED! 🥇 💯 
Great work on making the changes 👏 
You've done well implementing the requested changes

done

Your project is complete! There is nothing else to say other than... it's time to merge it :shipit:

## Highlights
- Great job, your project is error free :+1: 
- Nice work you meet all the requirements :1st_place_medal:

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Good luck! 🎉 


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: student.rb (hunk 1 above)
Comment (topeogunleye) [reply student.rb:18]: @Zven94, I apologize for the mistake. Yes, the play_hook method was required. 👍

**Pull request merged**